### PR TITLE
[develop][Performance] Split kv_b_proj and drop the strided copy before grouped matmul

### PR DESCRIPTION
**Pull request by DanielSun11**
### PR Category
Performance Optimization

### PR Types
Performance

### Description

latent MQA 路径上两处优化。

**1. 拆分 `kv_b_proj` 为独立吸收参数**

新增开关 `mqa_split_kv_b_proj`（仅 `hybrid_mla_attention` 为 `mqa_dsa` / `mqa_full_causal` 时有效）。开启后不再每次 forward 切片 `kv_b_proj.weight` 再走 `einsum`，改用两个已按 `fused_grouped_matmul` 布局存好的参数：`k_b_proj`（query 吸收）与 `v_b_proj`（V 去吸收），每侧一次 grouped GEMM，无切片、无转置。受 AOA 不能改 rank 限制，二者以折叠前两维的 2-D 存储，forward 内零拷贝 reshape 还原；`kv_b_proj` 在该模式下无梯度，缺这两个 key 的 checkpoint 由 loader（`_mla_split_kv_b_statements`）拆出。Muon 侧同步注册逐 head 切片（`axis=-2`），V 块转置回未拆分权重的朝向后再正交化，以保持各 `muon_version` 的缩放语义。

**2. `fused_grouped_matmul` 不再拷贝输入**

kernel 本身接收全部 stride，普通 strided view（如 `q[..., :qk_nope_head_dim]`）无需 `.contiguous()`。原实现每次 forward（含 recompute）及 backward 的 `dy` 都会白拷一次，而 phi 的 strided→dense kernel 无向量化特化（~350 GB/s vs dense ~4400 GB/s），代价与体积不成比例。新增 `_grouped_3d_strides` 只接受“dense 拷贝会逐元素线性化”的布局，broadcast / 负 stride / 自重叠 / 不可折叠一律回退原路径；入口另加 `w` 为 3-D `[G, R, D]`、`x.shape[-2:] == (G, D)` 的校验。

性能（`[1, 16384, 8, 128]` bf16）：fwd+bwd 1.080 → 0.869 ms/step（**1.24x**），`contiguous()` 调用 2/step → 0；8k profile 中约省 266 ms GPU 时间。

### 是否引起精度变化
否

零拷贝路径喂给同一 kernel 的值、顺序与 block 配置完全相同，结果**逐 bit 一致**：强制回退到拷贝路径后对比 `out` / 输入梯度 / `w.grad` 的 md5，312 种组合（13 种输入布局 × 2 dtype × 对齐与非对齐尺寸 × dense/非连续 `dy` × 三种冻结组合）与 3 个生产调用点全部零差异。拆分模式的等价性由 `TestSplitKvBProj` 覆盖：吸收 query、去吸收输出、整层 forward（含 warmup 的 eval/no-grad 早退路径）相对误差均 < 1e-3。

### CODE REVIEW

File: src/paddlefleet/transformer/transformer_config.py [hunk 1]
```diff
@@ -903,6 +903,32 @@ class TransformerConfig(ModelParallelConfig):
...
+    *stored* with the leading two dims folded, i.e. as 2-D
+    ``[heads * kv_lora_rank, qk_nope_head_dim]`` and
+    ``[heads * v_head_dim, kv_lora_rank]``; the 3-D form is recovered per forward
+    with a zero-copy reshape. The fold exists because the AOA engine cannot
+    change a tensor's rank, so the checkpoint side has to be 2-D. Each side is
+    then one grouped Triton GEMM with no slice, no einsum and no transpose.
+    ``kv_b_proj`` receives no gradient at all in this mode and the parameter set
+    is no longer byte-compatible with a dense MHA phase: a checkpoint without
+    these two keys must have them split out of ``kv_b_proj.weight`` by the loader
+    (``_mla_split_kv_b_statements`` generates the AOA statements that do it).
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/transformer_config.py]: P1 这里把旧检查点转换归因于 `_mla_split_kv_b_statements`，但当前可见的生产 loader/AOA 配置没有该实现或调用；新增测试的 `_loader_split`（`tests/single_card_tests/transformer/test_hybrid_mla_grad_health.py:462`）只是手工向 state dict 追加两个键。启用此开关加载仅含 `kv_b_proj.weight` 的旧检查点时，`k_b_proj`/`v_b_proj` 没有来源，会在 AOA/DCP 加载阶段缺失。请先落地实际的 AOA statements 及生产接入，并补一条从旧 state dict 经真实 loader 加载到目标模型的回归测试。

File: src/paddlefleet/triton_ops/grouped_matmul_fusion.py [hunk 2]
```diff
@@ -253,6 +253,45 @@ def grouped_matmul_dw_kernel(
...
+    # Innermost dim must be packed, and neither the group nor the row step may
+    # be degenerate (0 = broadcast, < 0 = reversed) or self-overlapping.
+    if stride_x != 1:
+        return None
+    if stride_g < shape[-1] or stride_m < shape[-2] * stride_g:
+        return None
+    # Leading dims collapse into M only if each is packed w.r.t. the next one.
+    # A size-1 dim contributes nothing to M, so its own stride is irrelevant.
+    for i in range(len(shape) - 3):
+        if shape[i] != 1 and strides[i] != shape[i + 1] * strides[i + 1]:
```
Comment (risemeup1111) [inline src/paddlefleet/triton_ops/grouped_matmul_fusion.py]: P1 这里跳过 singleton 维度时仍直接用该维度后面的 stride 校验前一轴，导致折叠后的行步长可能被错误接受。比如 shape `(2, 1, 3, 4, 6)`、strides `(100, 100, 24, 6, 1)` 会返回 `(24, 6, 1)`，但 6 个 M 行的真实起点是 `0,24,48,100,124,148`，kernel 按 `stride_m=24` 读取为 `0,24,48,72,96,120`，会产生错误结果甚至越界。请用跳过 singleton 后的有效折叠 stride 做连续性校验，或保守拒绝这类布局，并补充带非 unit 外层维度的 `as_strided` 回归测试。

File: src/paddlefleet/transformer/multi_latent_attention.py [hunk 3]
```diff
@@ -1464,6 +1491,48 @@ def __init__(
...
+            # they cannot produce a 3-D tensor. Exposing a reshaped *view* from
+            # ``sharded_state_dict`` instead is not safe: DCP reshapes the
+            # ``ShardedWeight.local_tensor`` in place (``reshape_`` /``flatten_``
+            # at ``dcp/load_state_dict.py:557,741,998``), which on a view that
+            # aliases the parameter corrupts the parameter's own shape.
+            #
+            # The 3-D form is recovered per forward with a zero-copy reshape.
+            # TP is forced to 1 in this mode, so ``num_attention_heads_per_partition``
+            # is the full head count and the parameter needs no TP attributes.
+            self.k_b_proj = self.create_parameter(
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/multi_latent_attention.py:1549]: P1 这个分支在保留 `kv_b_proj` 的同时又注册 `k_b_proj` 和 `v_b_proj`；两者的元素总数正好等于原 `kv_b_proj.weight`，而代码没有任何 state-dict/保存过滤（测试的 `_loader_split` 也只是把两个键追加到原字典）。因此开启 `mqa_split_kv_b_proj` 后该投影的常驻参数和检查点存储都会翻倍，且原参数在前向不再使用、不会更新，却仍会被保存，和 PR 描述的“parameter bytes are unchanged”相反。请改为唯一存储（例如让运行时权重引用拆分参数并调整保存映射），或明确移除旧参数并提供兼容迁移及参数/检查点大小回归测试。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
| --- | --- | --- | --- |
| 1 | [检查点转换](https://github.com/PaddlePaddle/PaddleFleet/pull/1696#discussion_r3749635151) | P1 | 🚧 |
| 2 | [分组矩阵乘 stride 校验](https://github.com/PaddlePaddle/PaddleFleet/pull/1696#discussion_r3749642080) | P1 | ✅ |
| 3 | [拆分参数存储](https://github.com/PaddlePaddle/PaddleFleet/pull/1696#discussion_r3749650109) | P1 | ✅ |
| 4 | [PR 描述](https://github.com/PaddlePaddle/PaddleFleet/pull/1696) | P3 | 🚧 |

P3 PR 描述已补充 `PR Category` 和 `PR Types`，但仍声称缺少 `k_b_proj` / `v_b_proj` 的 checkpoint 会由 `_mla_split_kv_b_statements` loader 拆出；该实现尚未接入，生产 `_gen_aoa_config` 仍只处理 `kv_b_proj.weight`。因此旧 checkpoint 转换的 P1 与描述准确性 P3 仍未解决。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/mqa_latent_attention.py [hunk 4]
```diff
@@ -353,7 +369,22 @@ def forward(
...
+        # turns a silently-wrong ``[G, R, D]`` / ``[l, h, v]`` mix-up (both
+        # reshape fine) into an error at the first forward.
+        contraction = v_b_proj_weight.shape[-1 if self.split_kv_b else 0]
+        if int(contraction) != kv_lora_rank:
+            raise ValueError(
+                "v_b_proj_weight layout mismatch: expected contraction dim "
+                f"kv_lora_rank={kv_lora_rank}, got shape "
+                f"{v_b_proj_weight.shape} with "
+                f"mqa_split_kv_b_proj={self.split_kv_b}."
+            )
```
[the remark this answers is not in the document]
Comment (DanielSun11) [reply src/paddlefleet/transformer/mqa_latent_attention.py:427]: 已修复：在读取 contraction 维之前先校验 rank，非 3-D 时直接报出带布局说明的 ValueError（`v_b_proj_weight must be 3-D [h, v_head_dim, kv_lora_rank] / [kv_lora_rank, h, v_head_dim]`），并补了 `test_v_b_proj_weight_rank_mismatch_rejected`。

File: src/paddlefleet/triton_ops/grouped_matmul_fusion.py (hunk 2 above)
Comment (DanielSun11) [reply src/paddlefleet/triton_ops/grouped_matmul_fusion.py]: 已修复：改为用“跳过 singleton 后的有效折叠范围（running span）”做连续性校验，外层维度必须等于其内部所有维度的实际跨度，span 再逐层左乘放大；singleton 只透传 span、不参与比较。你给的 shape `(2,1,3,4,6)` / strides `(100,100,24,6,1)` 现在返回 `None`，而 `(72,72,24,6,1)` 仍被接受为 `(24,6,1)`——两者都加进了 `test_rejects_unfoldable_dims_hidden_behind_a_singleton` 的 as_strided 回归测试。

File: src/paddlefleet/transformer/multi_latent_attention.py (hunk 3 above)
Comment (DanielSun11) [reply src/paddlefleet/transformer/multi_latent_attention.py:1549]: 已修复：split 模式下不再创建 `kv_b_proj`（`self.kv_b_proj = None`），`k_b_proj` + `v_b_proj` 的元素数正好等于原 `kv_b_proj.weight`，因此常驻参数与 checkpoint 大小都保持不变，也不会再保存一份不参与前向、永不更新的旧参数。配套改动：`_backward_kv_proj` 跳过 None；`create_parameter` 的 dtype 改用 `config.params_dtype`；由于 `enable_hy_sparse_attention` 会把层换成 `MQASelfAttention`（其 forward/decode 仍然读 `kv_b_proj.weight`），在 config 校验里显式拒绝该组合。测试：新增 `test_split_replaces_kv_b_proj_instead_of_duplicating_it`（断言 `kv_b_proj` 为 None、state_dict 无 `kv_b_proj.weight`、且 split 与 unsplit 的参数元素总数相等），`_loader_split` 改为 pop 掉旧 key，`test_split_kv_b_proj_rejects_hy_sparse_attention` 覆盖新校验。

File: src/paddlefleet/transformer/transformer_config.py (hunk 1 above)
Comment (DanielSun11) [reply src/paddlefleet/transformer/transformer_config.py]: 这条只做了文档修正，没有落地 AOA statements：`_mla_split_kv_b_statements` 在仓库里确实不存在（全仓 grep 只有这条 docstring 引用它），是我在描述里写早了。现已改为如实说明——启用该开关时 checkpoint 必须已经包含 `k_b_proj` / `v_b_proj`；从只含 `kv_b_proj.weight` 的旧 checkpoint 做转换所需的 AOA statements 尚未接入，因此这种 checkpoint 在该模式下无法直接续训。生产 loader 接入 + 真实 loader 的端到端回归测试超出本 PR 的范围（本 PR 只做前向/参数布局的性能改动），准备单独提一个 PR 来做，可以吗？

Comment (risemeup1111) [reply src/paddlefleet/transformer/transformer_config.py]: 暂不能关闭。该开关替换参数键后，既有训练/续训 checkpoint 仍只有 `kv_b_proj.weight`，当前生产 loader 无法生成 `k_b_proj` / `v_b_proj`；仅修改说明会把失败推迟到加载阶段。请在本 PR 接入实际 AOA/loader 转换及端到端回归，或暂不暴露该开关；此项仍保持 P1。

File: src/paddlefleet/triton_ops/grouped_matmul_fusion.py (hunk 2 above)
Comment (risemeup1111) [reply src/paddlefleet/triton_ops/grouped_matmul_fusion.py]: 已在当前 head 验证：实现改为按内部真实 span 逐层校验，原反例返回 `None`，对应可折叠布局仍返回 `(24, 6, 1)`，且新增 `as_strided` 回归覆盖。该项已解决。

File: src/paddlefleet/transformer/multi_latent_attention.py (hunk 3 above)
Comment (risemeup1111) [reply src/paddlefleet/transformer/multi_latent_attention.py:1549]: 已在当前 head 验证：split 分支不再构建 `kv_b_proj`，延迟反向路径会跳过该层，state dict 与参数总量回归也覆盖了唯一存储。该项已解决。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
旧 checkpoint 的生产 AOA/loader 转换仍未接入，现有 P1 继续阻塞。

**Review by From00 — APPROVED** [review]
LGTM
